File: Lab 5/Python/Task_1_Rayleigh.py
import numpy as np
from scipy.integrate import quad
import matplotlib.pyplot as plt
from sympy import *


def rayleigh_integral(sigma):
    PDF = lambda x: x*np.exp(-x**2/(2*sigma**2))/(sigma**2)
    total_prob, error_prob = quad(lambda x: x*np.exp(-x**2/(2*sigma**2))/(sigma**2), 0, np.inf)
    mean, error_m = quad(lambda x: x**2*np.exp(-x**2/(2*sigma**2))/(sigma**2), 0, np.inf)
    expectation_xsq_fx, error_var = quad(lambda x: x**3*np.exp(-x**2/(2*sigma**2))/(sigma**2), 0, np.inf)

    variance = expectation_xsq_fx - mean**2
    return [PDF, total_prob, mean, variance]


sigma = 1
xmin = 0
xmax = 5
fs = 10

x_plot = np.arange(xmin, xmax, 1/fs)
PDF, total_probability, mean, variance = rayleigh_integral(sigma)
print('Total Probability of the PDF', total_probability,
      '\nMean of the PDF', mean,
      '\nVariance of the PDF', variance)

plt.figure(1)
plt.plot(x_plot, PDF(x_plot))
plt.title('Rayleigh PDF')
plt.xlabel('Running Variable')
plt.ylabel('Amplitude')
plt.show()

# Additional -- Finding PDF using symbolic integration
x = Symbol('x')
PDF = x*exp(-x**2/(2*sigma**2))/(sigma**2)
CDF = integrate(PDF, x)
print('The CDF is', CDF)

File: Lab 5/MATLAB/Task_1_Rayleigh.m
% PDF of Rayleigh RV

sigma = sqrt(2);
fs = 10;
xmin = 0;
xmax = 5;

% Calculating probability, mean and variance using the integral formulae
% for CDF and expectation
[total_prob,mean_val,var_val] = RayleighRV(sigma,xmin,xmax,fs);
disp('The total probability (should be 1):')
disp(total_prob)
disp('Mean for the distribution:')
disp(mean_val)
disp('Variance for the distribution:')
disp(var_val)

function [q_0,mean_val,var_val] = RayleighRV(sigma,xmin,xmax,fs)
    x = xmin:1/fs:xmax;
    pdf = x.*exp(-(x).^2/(2*(sigma)^2))/(sigma^2);
    pdf_fn = @(t)t.*exp(-(t).^2/(2*(sigma)^2))/(sigma^2);
    exp_x_fn = @(t)t.^2.*exp(-(t).^2/(2*(sigma)^2))/(sigma^2);
    exp_xsq_fn = @(t)t.^3.*exp(-(t).^2/(2*(sigma)^2))/(sigma^2);

    figure()
    plot(x,pdf)

    title("Rayleigh RV")
    xlabel("Running variable")
    ylabel("PDF")
    grid on

    q_0 = integral(pdf_fn,0,inf);
    mean_val = integral(exp_x_fn,0,inf);
    var_val = integral(exp_xsq_fn,0,inf) - mean_val^2;
end